File: mcp/noctusai/tools/noctus/dev/ts_ast_edit.py
"""`noctus.dev.ts_ast_rename_symbol` — a REAL ts-morph AST edit for TypeScript.

WHY this tool exists
---------------------
Finding #7 (2026-07-23 session, memory `feedback_ast_first_rule_was_unenforced`):
the platform's AST-first rule ("if a compiler/type-checker parses the file,
edit it via libcst / ts-morph, never regex" — `KB § PATTERNS/common/ast.md`)
was **unenforceable for TypeScript** because no ts-morph tooling was
installed anywhere in the repo. `outline_typescript.py` deliberately stays
regex-based (read-only structural outline, no Node spawn — see its own
module docstring), so it never exercised the gap; every actual TS *edit*
in this repo has, until now, happened via `Edit`/string-replace, which
worked only because the diffs stayed small and safe by luck, not by
construction. This module is the first `noctus.dev.*` tool that performs
a genuine AST-driven TypeScript *write*.

WHERE ts-morph lives
---------------------
`mcp/noctusai/node/` — a **dedicated Node helper directory**, its own
`package.json` (`ts-morph` as the only dependency) + `node_modules/`
(gitignored, matches the existing global `node_modules` ignore rule).
Deliberately NOT:
  - a product's `frontend/` — that's product-runtime FE code, not
    dev-tooling; ts-morph is a devDependency of the *tooling*, not the
    shipped app.
  - `templates/product-seed/` — same reasoning; this isn't something a
    scaffolded product should inherit.
  - the MCP server's own `pyproject.toml`/`requirements.txt` — Node deps
    don't belong in a Python dependency manifest.
`outline_typescript.py`'s docstring already anticipated this shape
("Node spawn" as the thing it explicitly avoids for the *read-only* path)
— `mcp/noctusai/node/` is that spawn target for the *write* path.

HOW the edit happens
---------------------
`ts_ast_rename_symbol()` shells out to `node mcp/noctusai/node/ts_ast_edit.mjs`
(one JSON payload on stdin, one JSON result on stdout — see that file's
docstring for the op contract). The Node script loads the target file into
a real ts-morph `Project`, resolves the top-level declaration by name,
computes `findReferencesAsNodes()` BEFORE renaming (an honest count of what
actually changed — a regex `s/old/new/g` cannot report this truthfully:
it "succeeds" against 0 real references or over-matches unrelated
same-spelled text in strings/comments/JSX attrs), calls `.rename()`
(scope-aware — respects the real binder, not text matching), and writes
back via `SourceFile#saveSync()` (formatting-preserving).

Honest-error contract (no silent errors, `KB § 01-PHILOSOPHY.md`):
  - `node` not on PATH → `status="node_unavailable"`, `ok=False`. The
    caller (or a test) degrades explicitly; nothing silently no-ops.
  - Node script exits non-zero (a genuine crash, not a typed failure) →
    `status="node_crashed"`, stderr surfaced in `error`.
  - Node script's own typed failures (`parse_error`, `symbol_not_found`,
    `bad_request`, `unknown_op`, `rename_failed`, ...) pass through
    verbatim — the Python side does not re-interpret them.
"""
from __future__ import annotations

import json
import logging
import shutil
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

logger = logging.getLogger(__name__)

from settings import REPO_ROOT  # noqa: E402  (path constant)
from workspace import resolve_caller_root  # noqa: E402

# Per `feedback_mcp_path_constants_from_settings.md` — resolve off the
# `REPO_ROOT` settings constant, never `Path(__file__).parents[N]` (that
# broke 18 modules the last time the MCP toolkit's directory depth shifted).
_NODE_HELPER_DIR = REPO_ROOT / "mcp" / "noctusai" / "node"
_TS_AST_EDIT_SCRIPT = _NODE_HELPER_DIR / "ts_ast_edit.mjs"

_SUBPROCESS_TIMEOUT_SECONDS = 30


@dataclass
class TsAstEditResult:
    ok: bool
    status: str
    path: str
    old_name: Optional[str] = None
    new_name: Optional[str] = None
    references_updated: Optional[int] = None
    declaration_kind: Optional[str] = None
    error: Optional[str] = None

    def to_dict(self) -> dict:
        return {
            "ok": self.ok,
            "status": self.status,
            "path": self.path,
            "old_name": self.old_name,
            "new_name": self.new_name,
            "references_updated": self.references_updated,
            "declaration_kind": self.declaration_kind,
            "error": self.error,
        }


def _resolve_target_path(path: str, worktree_path: Optional[str]) -> Path:
    p = Path(path)
    if p.is_absolute():
        return p
    root = resolve_caller_root(worktree_path)
    return root / p


def _run_node_op(payload: dict) -> tuple[bool, dict, str]:
    """Run `ts_ast_edit.mjs` with `payload` on stdin.

    Returns ``(process_ok, result_dict, stderr)``. ``process_ok`` is False
    only for a genuine crash (non-zero exit / unparseable stdout) — a
    *typed* failure from the Node script (``ok: false`` in its own JSON)
    still returns ``process_ok=True`` with that JSON as ``result_dict``,
    since the subprocess itself ran fine and reported honestly.
    """
    proc = subprocess.run(
        ["node", str(_TS_AST_EDIT_SCRIPT)],
        input=json.dumps(payload),
        capture_output=True,
        text=True,
        timeout=_SUBPROCESS_TIMEOUT_SECONDS,
    )
    stderr = proc.stderr or ""
    if proc.returncode != 0:
        return False, {}, stderr
    try:
        result = json.loads(proc.stdout)
    except json.JSONDecodeError as exc:
        return False, {}, f"non-JSON stdout from ts_ast_edit.mjs: {exc}; stdout={proc.stdout!r}"
    return True, result, stderr


def ts_ast_rename_symbol(
    path: str,
    old_name: str,
    new_name: str,
    worktree_path: Optional[str] = None,
) -> TsAstEditResult:
    """Rename a top-level TS/TSX symbol AND every reference to it, via ts-morph.

    Covers function / class / interface / type-alias / const|let|var
    top-level declarations (the same set `outline_typescript.py` treats as
    "top-level structural declarations"). Scope-aware: a string literal,
    JSX attribute value, or comment that happens to contain the same text
    as ``old_name`` is left untouched — the case a `sed`/regex rename gets
    wrong (`KB § PATTERNS/common/ast.md § Anti-patterns`).
    """
    target = _resolve_target_path(path, worktree_path)

    if shutil.which("node") is None:
        msg = (
            "`node` not found on PATH — ts-morph AST edits require Node.js. "
            "This is a typed degrade, not a silent no-op: no file was touched."
        )
        logger.warning(msg)
        return TsAstEditResult(
            ok=False,
            status="node_unavailable",
            path=str(target),
            old_name=old_name,
            new_name=new_name,
            error=msg,
        )

    if not _TS_AST_EDIT_SCRIPT.exists():
        msg = f"ts_ast_edit.mjs missing at {_TS_AST_EDIT_SCRIPT} — node helper dir not installed correctly"
        logger.error(msg)
        return TsAstEditResult(
            ok=False,
            status="helper_script_missing",
            path=str(target),
            old_name=old_name,
            new_name=new_name,
            error=msg,
        )

    payload = {
        "op": "rename_symbol",
        "path": str(target),
        "oldName": old_name,
        "newName": new_name,
    }

    try:
        process_ok, result, stderr = _run_node_op(payload)
    except subprocess.TimeoutExpired:
        msg = f"ts_ast_edit.mjs timed out after {_SUBPROCESS_TIMEOUT_SECONDS}s on {target}"
        logger.error(msg)
        return TsAstEditResult(
            ok=False,
            status="timeout",
            path=str(target),
            old_name=old_name,
            new_name=new_name,
            error=msg,
        )

    if not process_ok:
        msg = f"ts_ast_edit.mjs crashed (non-JSON output or non-zero exit): {stderr.strip() or 'no stderr'}"
        logger.error(msg)
        return TsAstEditResult(
            ok=False,
            status="node_crashed",
            path=str(target),
            old_name=old_name,
            new_name=new_name,
            error=msg,
        )

    if not result.get("ok"):
        error = result.get("error", "unknown ts_ast_edit.mjs failure")
        logger.warning("ts_ast_rename_symbol failed: %s", error)
        return TsAstEditResult(
            ok=False,
            status=result.get("status", "unknown_error"),
            path=str(target),
            old_name=old_name,
            new_name=new_name,
            error=error,
        )

    return TsAstEditResult(
        ok=True,
        status=result.get("status", "renamed"),
        path=result.get("path", str(target)),
        old_name=result.get("oldName", old_name),
        new_name=result.get("newName", new_name),
        references_updated=result.get("referencesUpdated"),
        declaration_kind=result.get("declarationKind"),
    )


def register(server) -> None:
    @server.tool(
        name="noctus.dev.ts_ast_rename_symbol",
        description=(
            "Rename a top-level TS/TSX symbol (function/class/interface/type/"
            "const|let|var) AND every real reference to it, via a genuine "
            "ts-morph AST edit (scope-aware — resolves the TypeScript binder, "
            "not text matching). Leaves same-spelled string literals, JSX "
            "attribute values, and comments untouched — the case a regex/sed "
            "rename gets wrong. Honest typed error on missing node / parse "
            "failure / symbol-not-found (no silent no-op). "
            "KB § PATTERNS/common/ast.md."
        ),
    )
    def _ts_ast_rename_symbol(
        path: str,
        old_name: str,
        new_name: str,
        worktree_path: Optional[str] = None,
    ) -> dict:
        return ts_ast_rename_symbol(
            path=path,
            old_name=old_name,
            new_name=new_name,
            worktree_path=worktree_path,
        ).to_dict()

File: mcp/noctusai/node/ts_ast_edit.mjs
#!/usr/bin/env node
/**
 * ts_ast_edit.mjs — real ts-morph AST edits for `noctus.dev.ts_ast_*` MCP tools.
 *
 * Invoked by `mcp/noctusai/tools/noctus/dev/ts_ast_edit.py` as a child
 * process. Reads ONE JSON operation object from stdin, performs the AST
 * edit via ts-morph (the TypeScript Compiler API wrapper — see
 * `KB § PATTERNS/common/ast.md`), writes the result back to disk via
 * `SourceFile#saveSync()` (formatting-preserving), and prints ONE JSON
 * result object to stdout. Never touches stdout with anything but that
 * final JSON line — callers parse stdout as JSON.
 *
 * Supported ops (payload.op):
 *   - "rename_symbol": scope-aware rename of a top-level declared symbol
 *     (function / const / let / var / class / interface / type alias) and
 *     every reference to it *within this file*. This is the operation
 *     regex cannot do correctly: a same-spelled string literal, JSX prop
 *     value, comment, or unrelated identifier is left untouched because
 *     ts-morph resolves the identifier through the real binder, not text
 *     matching.
 *
 * Error contract: every failure path returns `{ok: false, status, error}`
 * on stdout with exit code 0 (the Python side distinguishes failure via
 * the `ok` field, not the process exit code — a genuine node crash /
 * unexpected exception still produces a non-zero exit + stderr, which the
 * Python side surfaces honestly as a "node_crashed" status). No silent
 * fallback to "did nothing" — an op that can't find its target reports a
 * typed error, it does not exit 0 with an empty edit.
 */
import { Project, SyntaxKind } from "ts-morph";

function readStdin() {
  return new Promise((resolve, reject) => {
    let data = "";
    process.stdin.setEncoding("utf8");
    process.stdin.on("data", (chunk) => {
      data += chunk;
    });
    process.stdin.on("end", () => resolve(data));
    process.stdin.on("error", reject);
  });
}

/** Find the top-level named declaration node for `name` in `sourceFile`.
 * Covers the declaration shapes `outline_typescript.py` already enumerates
 * as "top-level structural declarations": function / class / interface /
 * type alias / variable (const|let|var). Returns the ts-morph node that
 * exposes `.rename()`, or `null` if nothing top-level matches. */
function findRenameableDeclaration(sourceFile, name) {
  const fn = sourceFile.getFunction(name);
  if (fn) return fn;

  const cls = sourceFile.getClass(name);
  if (cls) return cls;

  const iface = sourceFile.getInterface(name);
  if (iface) return iface;

  const typeAlias = sourceFile.getTypeAlias(name);
  if (typeAlias) return typeAlias;

  const varDecl = sourceFile.getVariableDeclaration(name);
  if (varDecl) return varDecl;

  return null;
}

function opRenameSymbol(payload) {
  const { path: filePath, oldName, newName } = payload;
  if (!filePath || !oldName || !newName) {
    return {
      ok: false,
      status: "bad_request",
      error: "rename_symbol requires path, oldName, newName",
    };
  }

  const project = new Project({ useInMemoryFileSystem: false });
  let sourceFile;
  try {
    sourceFile = project.addSourceFileAtPath(filePath);
  } catch (err) {
    return {
      ok: false,
      status: "parse_error",
      error: `could not load/parse ${filePath}: ${String(err && err.message ? err.message : err)}`,
    };
  }

  const declaration = findRenameableDeclaration(sourceFile, oldName);
  if (!declaration) {
    return {
      ok: false,
      status: "symbol_not_found",
      error: `no top-level function/class/interface/type/const/let/var named "${oldName}" in ${filePath}`,
    };
  }

  // getNameNode() exists on all the declaration kinds above (VariableDeclaration,
  // FunctionDeclaration, ClassDeclaration, InterfaceDeclaration, TypeAliasDeclaration).
  const nameNode = typeof declaration.getNameNode === "function" ? declaration.getNameNode() : declaration;
  if (!nameNode) {
    return {
      ok: false,
      status: "symbol_not_renameable",
      error: `"${oldName}" was found but exposes no renameable name node`,
    };
  }

  // findReferencesAsNodes() BEFORE the rename gives us an honest count of
  // what actually changed — the thing a regex `s/old/new/g` cannot report
  // truthfully (it would "succeed" even against 0 real references, or over-
  // match unrelated same-spelled text).
  let referenceNodes;
  try {
    referenceNodes = nameNode.findReferencesAsNodes();
  } catch (err) {
    return {
      ok: false,
      status: "reference_resolution_failed",
      error: `findReferencesAsNodes failed for "${oldName}": ${String(err && err.message ? err.message : err)}`,
    };
  }

  try {
    nameNode.rename(newName);
  } catch (err) {
    return {
      ok: false,
      status: "rename_failed",
      error: `ts-morph rename("${newName}") failed: ${String(err && err.message ? err.message : err)}`,
    };
  }

  sourceFile.saveSync();

  return {
    ok: true,
    status: "renamed",
    path: filePath,
    oldName,
    newName,
    referencesUpdated: referenceNodes.length, // excludes the declaration itself; declaration renamed unconditionally
    declarationKind: SyntaxKind[declaration.getKind()],
  };
}

const OPS = {
  rename_symbol: opRenameSymbol,
};

async function main() {
  const raw = await readStdin();
  let payload;
  try {
    payload = JSON.parse(raw);
  } catch (err) {
    process.stdout.write(
      JSON.stringify({
        ok: false,
        status: "bad_payload",
        error: `stdin was not valid JSON: ${String(err && err.message ? err.message : err)}`,
      }),
    );
    return;
  }

  const op = OPS[payload.op];
  if (!op) {
    process.stdout.write(
      JSON.stringify({
        ok: false,
        status: "unknown_op",
        error: `unknown op "${payload.op}"; supported: ${Object.keys(OPS).join(", ")}`,
      }),
    );
    return;
  }

  const result = op(payload);
  process.stdout.write(JSON.stringify(result));
}

main().catch((err) => {
  // A genuine crash (bug in this script, ts-morph internal error, etc.) —
  // never swallow it. stderr + non-zero exit so the Python caller's
  // subprocess check surfaces it as "node_crashed", not a fake success.
  process.stderr.write(String(err && err.stack ? err.stack : err));
  process.exitCode = 1;
});
